packages/lib/src/utils/FocusLock.tsx

```diff
@@ -90,23 +90,43 @@ const useFocusableElements = (ref: React.MutableRefObject<HTMLDivElement>): HTML
 const FocusLock = ({ children }: { children: React.ReactNode }): JSX.Element => {
   const childrenContainerRef = useRef<HTMLDivElement>();
   const focusableElements = useFocusableElements(childrenContainerRef);
+  const initialFocus = useRef(false);
 
   const focusFirst = useCallback(() => {
     if (focusableElements?.length === 0) childrenContainerRef.current?.focus();
     else if (focusableElements?.length > 0) focusableElements.some((element) => attemptFocus(element));
   }, [focusableElements]);
 
   const focusLast = () => {
-    focusableElements?.reverse()?.some((element) => attemptFocus(element));
+    focusableElements?.slice().reverse()?.some((element) => attemptFocus(element));
   };
 
   const focusLock = (event: React.KeyboardEvent<HTMLDivElement>) => {
-    if (event.key === "Tab") focusableElements.length === 0 && event.preventDefault();
+    if (event.key === "Tab" && focusableElements.length === 0) event.preventDefault();
   };
 
   useEffect(() => {
-    if (!childrenContainerRef.current?.contains(document.activeElement) && !radixPortalContains(document.activeElement))
+    if (focusableElements !== undefined && !initialFocus.current) {
+      initialFocus.current = true;
       focusFirst();
+    }
+  }, [focusableElements, focusFirst]);
+
+  useEffect(() => {
+    const focusGuardHandler = (event: FocusEvent) => {
+      if (
+        !childrenContainerRef.current?.contains(event.relatedTarget as Node) &&
+        !childrenContainerRef.current?.nextElementSibling?.contains(event.relatedTarget as Node) &&
+        !childrenContainerRef.current?.previousElementSibling?.contains(event.relatedTarget as Node) &&
+        !radixPortalContains(event.relatedTarget as Node)
+      )
+        focusFirst();
+    };
+
+    document.addEventListener("focusout", focusGuardHandler);
+    return () => {
+      document.removeEventListener("focusout", focusGuardHandler);
+    };
   }, [focusFirst]);
```

Comment:
I think this should be refactored for easier readability and reduced code repetition:

```
useEffect(() => {
  const focusGuardHandler = (event: FocusEvent) => {
    const target = event.relatedTarget as Node | null;
    const container = childrenContainerRef.current;

    if (
      target &&
      !(
        container?.contains(target) ||
        container?.nextElementSibling?.contains(target) ||
        container?.previousElementSibling?.contains(target) ||
        radixPortalContains(target)
        // This can of course also be expressed as a conjunction 
        // of negations instead of negation of a disjunction, 
        // I find the latter a bit easier to comprehend.
      )
    ) {
      focusFirst();
    }
  };
  ```

Reply:
It's great Enrique, I will apply those changes.